### ISSUE TITLE
Documentation for `scatter` incorrectly states that index values must be unique

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#31776]
## 📚 Documentation
 
Documentation for `scatter_` and `scatter_add_` incorrectly states that "Moreover, as for gather(), the values of index must be between 0 and self.size(dim) - 1 inclusive, and all values in a row along the specified dimension dim must be unique."
https://pytorch.org/docs/stable/tensors.html?highlight=scatter_add#torch.Tensor.scatter_
There is no requirement for elements to be unique in either of 3 functions (`scatter_`, `scatter_add_`, `gather`), and `gather` documentation correctly does not mention that requirement.
In case of `scatter_`, there is no guarantees of which value from the source with the same indices ends up being written to `self`, in case of `scatter_add_` values from the source with the same indices are accumulated. 
There is a broader issue aside from the documentation here, because depending on whether indices are unique or repeating, parallel implementation for `scatter_add_` can be different (`scatter` is fine although can be non-deterministic). Maybe we should allow users to indicate if their indices are unique, so that we can choose more efficient implementation. The implementation variants are
1) indices are guaranteed to be unique - any naive implementation writing values from src to self will work, parallelization is easy
2) non-unique indices, but we are fine with non-deterministic behavior - we can use atomic accumulation in this case, and still do efficient parallelization
3) non-unique indices, and we want deterministic results - some more sophisticated algorithms required, handling accumulation for repeating indices (e.g. pre-sorting indices)
Currently we have 2) for cuda implementation and no parallelization for CPU, afaikt.
Setting "triage review" label to discuss desired behavior.
cc @nikitaved

**Comment by nikitaved** [comment]
Also, the current implementation of backward for `scatter/scatter_add` is only correct when `index.shape = source.shape` and indices are unique!

When `index.shape <= source.shape` and indices are not necessarily unique, the correct formula is:

```latex
Let
     I = {(i_1,...,i_n) | (i_1,...,i_n) < index.shape},
     K = {i in I | j:=(i_1,...,i_{d-1}, index[i], i_{d+1},...,i_n) = i, but self[j] != source[i] }
     U = {(i_1,...,i_n) | (i_1,...,i_n) < source.shape},
 co(I) = U/I,

source.grad[I] = grad.gather(d, index),
source.grad[col(I) union K] = 0,
```

where `K` is a set of indices which point to the same location in `self`, but did not contribute to it in forward `scatter/scatter_add`.

**Comment by nikitaved** [comment]
Also, it is not clear for `scatter_add` whether accumulation is a valid behavior. Accumulaton/single write require different formulas for backward.

**Comment by ngimel** [comment]
> Also, it is not clear for `scatter_add` whether accumulation is a valid behavior. Accumulaton/single write require different formulas for backward.

That's true, accumulation/single write require different formulas for backward, `gather` as a backward function is valid for accumulation case, because in that case all values from `src` contribute. 
Accumulation is valid and desirable behavior, because main use case for `scatter_add` is not in forward, but rather as a backward for `gather`, where accumulation is required to produce correct gradients.

**Comment by nikitaved** [comment]
All right, then the backward formula for `scatter_add` is then:
```latex
Let
     I = {(i_1,...,i_n) | (i_1,...,i_n) < index.shape},
     U = {(i_1,...,i_n) | (i_1,...,i_n) < source.shape},
 co(I) = U/I,

source.grad[I] = grad.gather(d, index),
source.grad[co(I)] = 0,
```

**Comment by ezyang** [comment]
@ngimel and I discussed this on Friday and we thought that a good first step is to improve the documentation. When you have non-unique indices, the backward will be wrong as it will propagate to all the places they could have come. So remove the warning that non-unique indices doesn't work (because it does work, and just arbitrarily picks one location), and warn that the gradient will be wrong. (Note that Numpy does the same thing in their documentation, we're following the best example here.)

**Comment by nikitaved** [comment]
Also, backward will throw an error when `index.shape < src.shape` because the current backward formula does not handle this case...
